pkg/k8s/resource_printer.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2020-Present VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package k8s
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	FormatYAML string = "yaml"
+	FormatJSON string = "json"
+)
+
+type ObjectPrinter interface {
+	PrintObject(obj runtime.Object, w io.Writer) error
+}
+
+func NewObjectPrinter(format string) (ObjectPrinter, error) {
+	switch format {
+	case FormatYAML:
+		return &YAMLObjectPrinter{}, nil
+	case FormatJSON:
+		return JSONObjectPrinter{}, nil
+	default:
+		return nil, fmt.Errorf("unsupported output format: %q, supported formats are yaml, json", format)
+	}
+}
+
+type YAMLObjectPrinter struct {
+	printCount int
+}
+
+func (y *YAMLObjectPrinter) PrintObject(obj runtime.Object, w io.Writer) error {
+	data, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	data, err = yaml.JSONToYAML(data)
+	if err != nil {
+		return err
+	}
+
+	y.printCount++
+	if y.printCount > 1 {
+		_, err := w.Write([]byte("---\n"))
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = w.Write(data)
+	return err
+}
+
+type JSONObjectPrinter struct{}
+
+func (j JSONObjectPrinter) PrintObject(obj runtime.Object, w io.Writer) error {
```

Comment:
What happens when this prints multiple objects?

Reply:
for YAML, multiple files printed out together could look like a single yaml file and hence the separator.

for JSON, the resource output is self contained, within the opening and closing braces {}, so did not put much thought into using a separator. Adding a `,` based separator would be easy. Happy to add them if you think it would be helpful.

Reply:
I just tried applying an output with 2 secrets in json format and they went through. So I think the json format doesn't need any form of separator?

```
$ cat create_secret.json
{
    "apiVersion": "v1",
    "metadata": {
        "name": "my-docker-hub-creds",
        "namespace": "default",
        "creationTimestamp": null
    },
    "kind": "Secret",
    "data": {
        ".dockerconfigjson": "redacted"
    },
    "type": "kubernetes.io/dockerconfigjson"
}
{
    "apiVersion": "v1",
    "metadata": {
        "name": "my-other-docker-hub-creds",
        "namespace": "default",
        "creationTimestamp": null
    },
    "kind": "Secret",
    "data": {
        ".dockerconfigjson": "redacted"
    },
    "type": "kubernetes.io/dockerconfigjson"
}
$ k apply -f create_secret.json
secret/my-docker-hub-creds configured
secret/my-other-docker-hub-creds created
```

Reply:
And kubectl does the [similar stuff](https://github.com/kubernetes/cli-runtime/blob/master/pkg/printers/json.go), based on which the ObjectPrinter was implemented

Reply:
Awesome!